option to reset offest automatically in case of OffsetOutOfRangeException

pjain1 opened the pull request:
if the next offset is less than the earliest available offset for that partition

pjain1 commented:
Follow up on https://github.com/druid-io/druid/issues/3467

himanshug commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java:
```diff
@@ -386,16 +385,58 @@ public void run()
+                  }
+                }
+                // check if we seeked passed the endOffset for this partition
+                if (consumer.position(topicPartitions) >= endOffsets.get(topicPartitions.partition())
```
i believe, we can move this if block in the above if block itself.

pjain1 replied:
done. thanks

himanshug commented:
:+1: besides the minor comment

dclim commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java:
```diff
@@ -386,16 +385,58 @@ public void run()
+                  } else {
+                    log.makeAlert("Got OffsetOutOfRangeException [%s], resetting to the latest offset", e.getMessage());
+                    consumer.seekToEnd(topicPartition);
+                    nextOffsets.put(topicPartition.partition(), consumer.position(topicPartition));
```
These lines are repeated and can be pulled out of the `if (ioConfig.isUseEarliestOffset())`

pjain1 replied:
done

dclim commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java:
```diff
@@ -386,16 +385,58 @@ public void run()
+            // Reset consumer offset if resetOffsetAutomatically is set to true
+            // and the current message offset in the kafka partition is more than the
+            // next message offset that we are trying to fetch
+            if (tuningConfig.isResetOffsetAutomatically()) {
```
Can all this logic be moved into a method to keep the main while loop as clean as possible?

pjain1 replied:
done

dclim commented on docs/content/development/extensions-core/kafka-ingestion.md:
```diff
@@ -123,6 +123,7 @@ The tuningConfig is optional and default parameters will be used if no tuningCon
 |`buildV9Directly`|Boolean|Whether to build a v9 index directly instead of first building a v8 index and then converting it to v9 format.|no (default == false)|
 |`reportParseExceptions`|Boolean|If true, exceptions encountered during parsing will be thrown and will halt ingestion; if false, unparseable rows and fields will be skipped.|no (default == false)|
 |`handoffConditionTimeout`|Long|Milliseconds to wait for segment handoff. It must be >= 0, where 0 means to wait forever.|no (default == 0)|
+|`resetOffsetAutomatically`|Boolean|Whether to reset the consumer offset if the next offset that it is trying to fetch is less than the earliest available offset for that particular partition. The consumer offset will be reset to either the earliest or latest offset depending on `useEarliestOffset` property of `KafkaSupervisorIOConfig` (see below). WARNING - Getting into this situation generally  ...
```
Instead of saying 'leads to data loss' which could be misconstrued to mean they'll lose existing data in Druid, maybe phrase it something more like 'this situation typically occurs when messages in Kafka are no longer available for consumption and therefore won't be ingested into Druid'

dclim commented:
Are you able to add tests for this functionality?

pjain1 commented:
@dclim I refactored the code a little bit and took care of review comments, please have a look now. I added a unit test where next offset to be fetched is greater than the least available offset which means no resetting should happen. I could not add a UT that tests the reset as it hard to simulate data drop from kafka in UT, I tried setting the retention to a very small value for the TestBroker but did not work, please let me know if you any other idea to test this. Also, we are running this code in our metrics cluster on one of the MMs.

pjain1 commented:
@himanshug can you have a look again ?

himanshug commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java:
```diff
@@ -1000,4 +991,74 @@ private boolean possiblyPause(Set<Integer> assignment) throws InterruptedExcepti
+  ) throws InterruptedException
+  {
+    boolean shouldRetry = false;
+    for (Map.Entry<TopicPartition, Long> outOfRangePartition : outOfRangePartitions.entrySet()) {
```
why do we need to run this code when `resetOffsetAutomatically ` is set to false.

pjain1 replied:
there is no need but I thought it is easier to read this way than having nested if else for checking `tuningConfig.isResetOffsetAutomatically` and `leastAvailableOffset > nextOffset` and setting `shouldRetry` at at multiple places

pjain1 replied:
anyways refactored the code to skip the for loop in case reset automatic option is not enabled

himanshug commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java:
```diff
@@ -386,17 +385,9 @@ public void run()
-            }
+            log.warn("OffsetOutOfRangeException with message [%s]", e.getMessage());
+            possiblyResetOffsetsOrWait(e.offsetOutOfRangePartitions(), consumer, assignment);
+            stillReading = ioConfig.isPauseAfterRead() || !assignment.isEmpty();
```
this line did not exist before. so, I'm assuming this is only needed when `resetOffsetAutomatically ` is set to true, why then do we have it for both cases now?

pjain1 replied:
yes it was not there but it should not cause any problem @dclim what do you think ?

pjain1 commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java:
```diff
@@ -1000,4 +991,78 @@ private boolean possiblyPause(Set<Integer> assignment) throws InterruptedExcepti
+      consumer.seekToEnd(topicPartition);
+    }
+    nextOffsets.put(topicPartition.partition(), consumer.position(topicPartition));
+    log.warn("Consumer is now at offset [%d]", consumer.position(topicPartition));
```
thanks.. done

dclim commented:
Hm, I tried simulating offsets being evicted in a UT friendly way and also wasn't successful. I tested it manually and it looks good.

dclim commented:
👍 LGTM

himanshug commented:
:+1: @dclim thanks for doing the extra testing
